orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -188,6 +192,13 @@ def forward(self, *inputs, **kwargs):
             self._forward_session = onnxruntime.InferenceSession(self._onnx_forward.SerializeToString())
             self._backward_session = onnxruntime.InferenceSession(self._onnx_backward.SerializeToString())
 
+            if self._device.type == 'cuda':
```

Comment:
As per documentation, all provders should be used when it is not specified. Maybe we have a bug on `InferenceSession` that we should fix instead:

```
class InferenceSession(Session):
    """
    This is the main class used to run a model.
    """
    def __init__(self, path_or_bytes, sess_options=None, providers=None, provider_options=None):
        """
        :param path_or_bytes: filename or serialized ONNX or ORT format model in a byte string
        :param sess_options: session options
        :param providers: list of providers to use for session. If empty, will use all available providers.
        :param provider_options: list of provider options dict for each provider, in the same order as 'providers'

        The model type will be inferred unless explicitly set in the SessionOptions.
        To explicitly set:
          so = onnxruntime.SessionOptions()
          so.add_session_config_entry('session.load_model_format', 'ONNX') or
          so.add_session_config_entry('session.load_model_format', 'ORT') or

        A file extension of '.ort' will be inferred as an ORT format model.
        All other filenames are assumed to be ONNX format models.
        """
```

Reply:
What's the bug? The documentation on InferenceSession doesn't constrain what we do in ORTModule.

Either way, we only have training kernels registered in the CPU and CUDA providers, so I'm not sure what the effect would be of opening this up to other providers....